Packages/ClientRuntime/Sources/Waiters/WaiterIterator.swift

```diff
@@ -0,0 +1,86 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// `WaiterIterator` performs the waited operation using the timing in the retry strategy.
+/// A new iterator is created for every wait operation.
+/// Once the wait operation concludes with either success or failure, the `WaiterIterator`
+/// ends iteration.
+public class WaiterIterator<Input, Output>: AsyncIteratorProtocol {
+    let input: Input
+    let acceptors: [Acceptor<Input, Output>]
+    let scheduler: WaiterScheduler
+    let operation: (Input) async throws -> Output
+
+    private var finished = false
+
+    init(input: Input, acceptors: [Acceptor<Input, Output>], minDelay: TimeInterval, maxDelay: TimeInterval, maximumWaitTime: TimeInterval, operation: @escaping (Input) async throws -> Output) {
```

Comment:
Nit: Can we put arguments on new lines when we have more than ~3 or 4? (Perhaps we should add a linter rule for this

```suggestion
    init(
        input: Input, 
        acceptors: [Acceptor<Input, Output>], 
        minDelay: TimeInterval, 
        maxDelay: TimeInterval, 
        maximumWaitTime: TimeInterval, 
        operation: @escaping (Input) async throws -> Output
    ) {
```